src/accountsdb/cache.zig

```diff
@@ -0,0 +1,331 @@
+            }
+        };
+
+        if (self.highest_slot == null or slot > self.highest_slot.?) {
+            self.highest_slot = slot;
+        }
+
+        self.enforceMaxSlotCount();
+
+        const new_cached_account = try self.allocator.create(CachedAccount);
+        errdefer self.allocator.destroy(new_cached_account);
+        new_cached_account.* = try CachedAccount.init(self.allocator, account);
+        errdefer new_cached_account.deinit(self.allocator);
+
+        slot_lru.putNoClobber(pubkey, new_cached_account) catch |err| switch (err) {
+            error.EntryAlreadyExists => return error.AlreadyExistsInCache,
+        };
+    }
+
+    /// remove slot lru, decreasing ref_counts on CachedAccounts (optionally removing)
+    pub fn purgeSlot(self: *Self, slot: Slot) void {
+        const slot_lru = self.slot_lrus.getPtr(slot) orelse return;
+        slot_lru.deinit();
+        _ = self.slot_lrus.remove(slot);
+    }
+
+    /// bring slot lru forward to new slot, increasing ref_counts
+    pub fn shiftLRUForward(self: *Self, old_slot: Slot, new_slot: Slot) !void {
+        const old_slot_lru = self.slot_lrus.getPtr(old_slot) orelse return error.SlotNotFound;
+
+        var new_slot_lru = try LRU.initWithContext(self.allocator, self.max_items, self.allocator);
+
+        // copy from old to new slot lru
+        {
+            old_slot_lru.mux.lock();
+            defer old_slot_lru.mux.unlock();
+
+            var it = old_slot_lru.dbl_link_list.first;
+            while (it) |node| : (it = node.next) {
+                if (node.data.value.is_dirty.load(.acquire)) continue; // do not copy forward modified accounts
+
+                _ = new_slot_lru.put(node.data.key, node.data.value.copyRef());
+            }
+        }
+
+        try self.slot_lrus.put(self.allocator, new_slot, new_slot_lru);
+
+        self.enforceMaxSlotCount();
+
+        if (self.highest_slot == null or new_slot > self.highest_slot.?) {
+            self.highest_slot = new_slot;
+        }
+    }
+
+    /// removes lowest slot when exceeding .max_slots
+    pub fn enforceMaxSlotCount(self: *Self) void {
+        if (self.slot_lrus.count() > self.max_slots) {
+            var lowest_slot: Slot = std.math.maxInt(usize);
+            var iter = self.slot_lrus.iterator();
+            while (iter.next()) |entry| {
+                if (entry.key_ptr.* < lowest_slot) lowest_slot = entry.key_ptr.*;
+            }
+            self.purgeSlot(lowest_slot);
+        }
+    }
+
+    pub fn deinit(self: *Self) void {
+        var slot_iter = self.slot_lrus.iterator();
+        while (slot_iter.next()) |entry| {
+            const slot_lru_ptr = entry.value_ptr;
+            slot_lru_ptr.deinit();
+        }
+        self.slot_lrus.deinit(self.allocator);
+    }
+};
+
+test "CachedAccount ref_count" {
+    const allocator = std.testing.allocator;
+    var random = std.rand.DefaultPrng.init(19);
+    const rng = random.random();
```

Comment:
To this and all other instances of this in the other tests: I think it would be worth falling in line with #304, which looks like it's going to be merged.